source/routing/journey_pattern_container.cpp

```diff
@@ -0,0 +1,160 @@
+        for (const auto& vj: jp->frequency_vehicle_journey_list) { add_vj(*vj); }
+    }
+}
+
+JourneyPatternContainer::JpKey JourneyPatternContainer::make_key(const nt::VehicleJourney& vj) {
+    JourneyPatternContainer::JpKey key;
+    key.route_idx = RouteIdx(*vj.journey_pattern->route);
+    key.is_freq = dynamic_cast<const nt::FrequencyVehicleJourney*>(&vj) != nullptr;
+    for (const auto& st: vj.stop_time_list) {
+        key.jpp_keys.emplace_back(SpIdx(*st.journey_pattern_point->stop_point),
+                                  st.local_traffic_zone,
+                                  st.properties);
+    }
+    return key;
+}
+
+static bool st_are_equals(const nt::VehicleJourney& vj1, const nt::VehicleJourney& vj2) {
+    for (auto it1 = vj1.stop_time_list.begin(), it2 = vj2.stop_time_list.begin();
+         it1 !=  vj1.stop_time_list.end() && it2 !=  vj2.stop_time_list.end();
+         ++it1, ++it2) {
+        if (it1->arrival_time != it2->arrival_time || it1->departure_time != it2->departure_time) {
+            return false;
+        }
+    }
+    return true;
+}
+
+template<typename VJ> static bool
+overtake(const VJ& vj, const std::vector<const VJ*> vjs) {
+    if (vj.stop_time_list.empty()) { return false; }
+    for (const VJ* cur_vj: vjs) {
+        assert(vj.stop_time_list.size() == cur_vj->stop_time_list.size());
+
+        // if the validity patterns do not overlap, it can't overtake
+        if ((vj.validity_pattern->days & cur_vj->validity_pattern->days).none() &&
+            (vj.adapted_validity_pattern->days & cur_vj->adapted_validity_pattern->days).none()) {
+            continue;
+        }
+
+        // if the stop times are the same, they don't overtake
+        if (st_are_equals(vj, *cur_vj)) { continue; }
+
+        const bool vj_is_first =
+            vj.stop_time_list.front().departure_time < cur_vj->stop_time_list.front().departure_time;
+        const VJ& vj1 = vj_is_first ? vj : *cur_vj;
+        const VJ& vj2 = vj_is_first ? *cur_vj : vj;
+        for (auto it1 = vj1.stop_time_list.begin(), it2 = vj2.stop_time_list.begin();
+             it1 !=  vj1.stop_time_list.end() && it2 !=  vj2.stop_time_list.end();
+             ++it1, ++it2) {
+            if (it1->arrival_time >= it2->arrival_time ||
+                it1->departure_time >= it2->departure_time) {
+                return true;
+            }            
+        }
+    }
+    return false;
+}
+
+template<typename VJ>
+void JourneyPatternContainer::add_vj(const VJ& vj) {
+    // Get the existing jps according to the key.
+    const auto key = make_key(vj);
+    auto& jps = map[key];
+
+    // The vj must not overtake the other vj in its jp, thus searching
+    // for one satisfying jp.
+    for (auto& jp_idx: jps) {
+        auto& jp = get(jp_idx);
+        auto& vjs = jp.template get_vjs<VJ>();
+        if (! overtake(vj, vjs)) {
+            // Great, found a valid jp, inserting and stopping
```

Comment:
nop, curent journey pattern is valid as our vj does not overtake the vjs in the jp.